server/models.py

```diff
@@ -607,6 +616,13 @@ def _unflag_all(self, member_ids):
         if backup:
             backup.flagged = False
 
+    @transaction
+    def publish_grades(self, tag):
+        """Publish student grades for the assignment."""
+        self.grade_visibility = tag
+        db.session.add(self)
```

Comment:
I don't think this method wants to do an add - it is not inserting an row here. You also don't need `db.session.commit()` because the decorator will do it for you.